test/e2e/minimal_configuration.go

```diff
@@ -251,5 +278,146 @@ var _ = Describe("Create providers with minimal specified configuration", func()
+			key := client.ObjectKey{Namespace: operatorNamespace, Name: coreProviderName}
+			if err := k8sclient.Get(ctx, key, coreProvider); err != nil {
+				return false
+			}
+
+			if coreProvider.Status.InstalledVersion != nil && *coreProvider.Status.InstalledVersion == previousCAPIVersion {
+				return true
+			}
+			return false
+		}, timeout).Should(Equal(true))
+	})
+
+	It("should successfully upgrade a CoreProvider (v1.4.0 -> v1.4.2)", func() {
+		k8sclient := bootstrapClusterProxy.GetClient()
+		coreProvider := &operatorv1.CoreProvider{}
+		key := client.ObjectKey{Namespace: operatorNamespace, Name: coreProviderName}
+		Expect(k8sclient.Get(ctx, key, coreProvider)).To(Succeed())
+
+		coreProvider.Spec.Version = capiVersion
+
+		Expect(k8sclient.Update(ctx, coreProvider)).To(Succeed())
+
+		By("Waiting for the core provider deployment to be ready")
+		Eventually(func() bool {
+			isReady, err := waitForDeployment(k8sclient, ctx, coreProviderDeploymentName)
+			if err != nil {
+				return false
+			}
+			return isReady
+		}, timeout).Should(Equal(true))
+
+		By("Waiting for core provider to be ready")
+		Eventually(func() bool {
+			coreProvider := &operatorv1.CoreProvider{}
+			key := client.ObjectKey{Namespace: operatorNamespace, Name: coreProviderName}
+			if err := k8sclient.Get(ctx, key, coreProvider); err != nil {
+				return false
+			}
+
+			for _, c := range coreProvider.Status.Conditions {
+				if c.Type == operatorv1.ProviderInstalledCondition && c.Status == corev1.ConditionTrue {
+					return true
+				}
+			}
+			return false
+		}, timeout).Should(Equal(true))
+
+		By("Waiting for status.IntalledVersion to be set")
+		Eventually(func() bool {
+			coreProvider := &operatorv1.CoreProvider{}
+			key := client.ObjectKey{Namespace: operatorNamespace, Name: coreProviderName}
+			if err := k8sclient.Get(ctx, key, coreProvider); err != nil {
+				return false
+			}
+
+			if coreProvider.Status.InstalledVersion != nil && *coreProvider.Status.InstalledVersion == capiVersion {
+				return true
+			}
+			return false
+		}, timeout).Should(Equal(true))
+	})
+
+	It("should successfully delete a CoreProvider", func() {
+		k8sclient := bootstrapClusterProxy.GetClient()
+		coreProvider := &operatorv1.CoreProvider{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      coreProviderName,
+				Namespace: operatorNamespace,
+			},
+			Spec: operatorv1.CoreProviderSpec{
+				ProviderSpec: operatorv1.ProviderSpec{
+					Version: capiVersion,
+				},
+			},
+		}
+
+		Expect(k8sclient.Delete(ctx, coreProvider)).To(Succeed())
+
+		By("Waiting for the core provider deployment to be deleted")
+		Eventually(func() bool {
+			deployment := &appsv1.Deployment{}
+			key := client.ObjectKey{Namespace: operatorNamespace, Name: coreProviderDeploymentName}
+			isReady, err := waitForObjectToBeDeleted(k8sclient, ctx, key, deployment)
+			if err != nil {
+				return false
+			}
+			return isReady
+		}, timeout).Should(Equal(true))
```

Comment:
perhaps it is better if we will move this code to the end of core provider tests at the beginning of the file and drop the dedicated "It" test case to shorten the code.

Reply:
Unfortunately we can't delete it earlier, because other providers will fail a preflight check that tests that Core Provider is Ready: https://github.com/kubernetes-sigs/cluster-api-operator/blob/main/internal/controller/preflight_checks.go#L130-L131